sympy/core/tests/test_match.py

```diff
@@ -134,7 +134,7 @@ def test_mul():
     assert e.match(x**p*exp(x*q)) == {p: 0, q: 1}
 
     e = I*Poly(x, x)
-    assert e.match(I*p) == {p: Poly(x, x)}
+    assert e.match(I*p) == {p: x}
```

Comment:
This seems to be a regression.

Reply:
The new behavior seems just as correct to me. `match` is only defined on Poly because it subclasses from Basic, and it fakes `.args`, giving this result. The previous behavior of `Poly` being in a `Mul` was even more incorrect, I'd say. 

I'm actually questioning whether making Poly Basic was a good idea. I'm thinking it wasn't. If it wasn't Basic, I think `__rmul__` would be called automatically.  But a change like that would be a bit of a compatibility break, so I'm not sure if we can do it.